### PR TITLE
Only apply one document info change for linked documents

### PR DESCRIPTION
**Pull request by Cosifne**
Fix https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1903953/
So the story is, when renaming a class (also rename the containing type) in multiple TFMs (let's say two TFMs), there would be
1. Document change.
2. Document info change
for each of the linked documents.

**In current Implementation**
We will apply the document change and document info (rename file) change for the first documents. And the `info change` will call this [line ](https://github.com/dotnet/roslyn/blob/8c38000b3bb2fb64699633eb58e0d284cb3a0ed1/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs#L1212). Start from here we are asking shell -> project system -> our project system listeners to update the document in `Workspace.CurrentSolution`. As a result, both of the linked documents will be updated.
And when Workspace tries to apply the change for the second document, it can't find it (because it get changed because of the first document rename) therefore the rename operation just fails.

**Changes in this PR**
Separate the info change into a sperate method and do that only after applying project changes. Also, for linked document, we only need to apply one info change since it's going to rename all the linked document from the file system.

~TODO:
I want to add test coverage. But right now the integration test harness is broken so tests can't be launched from VS. Will need to wait it to be resolved first.~ (Done)

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:1483]: This seems like it could break things that should work.  Say, for example, the user performs a fix-all on code like:

```c#
#if DEBUG

// code that is active in some project

#else

// other code that is not.

#endif
```

with the normal batch fixer, this would work, right?  Or maybe it wouldn't?  I'm not exactly sure.

Comment (Cosifne) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:1483]: I feel it won't be affected. This change would only affect things like file name, path change. For content change, each document would still be changed even they are linked.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:1483]: good ot know!  probably worth a test :)

**Comment by Cosifne** [comment]
OK, tag @jasonmalinowski  to review.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:1483]: I imagine we might end up actually taking the ApplyChangedDocument path more than once, but we'll end up writing the same text each time -- the documents should have had merged texts. If we look under a debugger and see we are indeed doing that, that might be a nice potential win here to only process that once.

That said, I'm not sure how we exactly fit that in here -- we're just calling ApplyProjectChanges for both projects and we can't just pass additional parameters since that's a protected API. So either we have to break the API, or maybe pre-filter the ProjectChanges to remove the duplicate changes.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:1473]: Since you've already got this list, consider just adding all of them in line 1478. That then removes the need for the .All() check and you can just check the Id on line 1471 before doing anything further. The idea being rather than each time we process re-checking if this is a known duplicate, just add to the hash set the document Ids we've already "covered". Might consume a hair more memory but would reduce overall work.

**Comment by jasonmalinowski** [comment]
@Cosifne Definitely a good start, let's get integration tests working before merge. @CyrusNajmabadi's point about duplicates does make me think there might be a nice win too around applying text edits that we might want to do as well, if you can make the code work without breaking APIs too much.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs
Comment (Cosifne) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:1481]: So, it sounds like a future work? (e.g. we have to see how things work for text changes)

**Comment by Cosifne** [comment]
@jasonmalinowski  Review for second review

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpRename.cs
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpRename.cs:735]:
```suggestion
    <LangVersion>preview</LangVersion>
```

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:1476]:
```suggestion
                        var linkedDocumentIds = oldDoc.GetLinkedDocumentIds();
```

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**
